### ISSUE TITLE
Row not fully deleted from cassandra on em.remove(obj) - then cannot reinsert row with same key

### ISSUE THREAD

**Issue by masha256** [Impetus/kundera#130]
I have come across a strange issue when removing a row from cassandra and then trying to re-insert a row with the same row key. Once the row is deleted, if you re-persist a new object with the same row key, it looks like the object is persisted, but its not really put into cassandra but it is find()'able during that JVM run (perhaps its in the cache?).

Here is a test case that can be used:

It needs to be run first with firstRun set to true and then ran again with firstRun set to false and then finally run again with firstRun set to true. The reason it needs to be separate runs is I am guessing the cache is involved.

``` java
package com.walmartlabs.dal.kundera.test;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.Persistence;

import org.junit.Test;

import com.walmartlabs.dal.kundera.test.dao.User;

public class KunderaRowDeleteTest {

    @Test
    public void test() {


        boolean firstRun = true;


        EntityManagerFactory emf = Persistence.createEntityManagerFactory("cassandra_pu");
        EntityManager em = emf.createEntityManager();



        if (firstRun)
        {
            User user = new User();
            user.setUserId("0001");
            user.setFirstName("John");
            user.setLastName("Smith");
            user.setCity("London");

            em.persist(user);
        }
        else
        {
            User found = em.find(User.class, "0001");
            if (found != null)
            {
                System.out.println("Removing found user " + found.getFirstName());
                em.remove(found);
            } else
            {
                System.out.println("User not found");
            }
        }


        em.close();
        emf.close();
    }

}
```

Entity:

``` java
package com.walmartlabs.dal.kundera.test.dao;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.Table;

@Entity
@Table(name = "users", schema = "KunderaExamples@cassandra_pu")
public class User 
{
    @Id
    private String userId;

    @Column(name="first_name")
    private String firstName;

    @Column(name="last_name")
    private String lastName;

    @Column(name="city")
    private String city;    

    public User()
    {
    }

    public String getUserId()
    {
        return userId;
    }
    public void setUserId(String userId)
    {
        this.userId = userId;
    }
    public String getFirstName()
    {
        return firstName;
    }
    public void setFirstName(String firstName)
    {
        this.firstName = firstName;
    }
    public String getLastName()
    {
        return lastName;
    }
    public void setLastName(String lastName)
    {
        this.lastName = lastName;
    }
    public String getCity()
    {
        return city;
    }
    public void setCity(String city)
    {
        this.city = city;
    } 
}
```

persistence.xml

``` xml
<persistence xmlns="http://java.sun.com/xml/ns/persistence" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://java.sun.com/xml/ns/persistence
    http://java.sun.com/xml/ns/persistence/persistence_2_0.xsd"
    version="2.0">
    <persistence-unit name="cassandra_pu">
        <provider>com.impetus.kundera.KunderaPersistence</provider>     
        <properties>            
            <property name="kundera.nodes" value="localhost"/>
            <property name="kundera.port" value="9160"/>
            <property name="kundera.keyspace" value="KunderaExamples"/>
            <property name="kundera.dialect" value="cassandra"/>
            <property name="kundera.client.lookup.class" value="com.impetus.client.cassandra.pelops.PelopsClientFactory" />
            <property name="kundera.cache.provider.class" value="com.impetus.kundera.cache.ehcache.EhCacheProvider"/>
            <property name="kundera.cache.config.resource" value="/ehcache-test.xml"/>
        </properties>       
    </persistence-unit>

</persistence>

```

After the second run (one with firstRun = true and once with firstRun = false), this is what shows up in the cassandra-cli:

<pre>
[default@KunderaExamples] list users;
Using default limit of 100
-------------------
RowKey: 0001

1 Row Returned.
Elapsed time: 9 msec(s).
</pre>

Even if you run the code again with firstRun = true, it should re-persist the object, if you run another cli 'list users' you still do not see the object. The only time you see the full object existing in cassandra on the cli is after the very first run of the above code (or after issuing 'truncate users' on the cli and running the firstRun=true again):

<pre>
[default@KunderaExamples] list users;
Using default limit of 100
-------------------
RowKey: 0001
=> (column=63697479, value=London, timestamp=1349911254602)
=> (column=66697273745f6e616d65, value=John, timestamp=1349911254602)
=> (column=6c6173745f6e616d65, value=Smith, timestamp=1349911254602)

1 Row Returned.
Elapsed time: 16 msec(s).

</pre>

**Comment by masha256** [comment]
FYI - I was looking at the insert after delete unit test that was added a couple days ago, and I think that test is on the right track but missed something that would show this problem (at least in my case). In the StudentCassandraIntTest unit test: testDelete() calls testPersist() to see if it can insert after the delete, but it is never verified that the new insert actually inserted anything.

I suggest adding a call to testFindById() directly after the testPersist() call in testDelete(). I believe that would show the rows are not actually getting inserted after the delete case.

**Comment by mevivs** [comment]
This should work with :

```
                        <property name="kundera.client.lookup.class" value="com.impetus.client.cassandra.thrift.ThriftClientFactory" />
```

It is a problem with PelopsClientFactory but not with ThriftClientFactory.

**Comment by masha256** [comment]
Yep that fixed it. Any reason that is not the default lookup class in the examples/documentation?

**Comment by mevivs** [comment]
Since it's inception Kundera relied on Pelops API to interact with cassandra. However due to less upgrade and support we build own Thriftclient and released with 2.1 . If you have a look in current trunk branch, default client factory lookup is now "ThriftClientFactory".

-Vivek
